golang_src/cmd/worker.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"crypto/rand"
+	"fmt"
+	"github.com/cerberus98/babymailgun"
+	"github.com/spf13/viper"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+type WorkerConfig struct {
+	Host         string
+	Port         string
+	DatabaseName string
+	WorkerSleep  int
+	WorkerPool   int
+	MailHost     string
+	MailPort     string
+	AdminEmail   string
+}
+
+// Create a UUID4. Source implementation here: https://groups.google.com/forum/#!msg/golang-nuts/d0nF_k4dSx4/rPGgfXv6QCoJ
+func uuid() string {
+	b := make([]byte, 16)
+	rand.Read(b)
+	b[6] = (b[6] & 0x0f) | 0x40
+	b[8] = (b[8] & 0x3f) | 0x80
+	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
+}
+
+func processingLoop(cfg *WorkerConfig, wg *sync.WaitGroup, quit <-chan bool) {
+	// Fetch emails ready to be sent
+	// Try to send them
+	// Update the datastore
+	// Go back to sleep
+	log.Println("Starting processing loop...")
+	workerId := uuid()
+	defer wg.Done()
+loop:
+	for {
+		mongoClient := babymailgun.MongoClient{Host: cfg.Host, Port: cfg.Port, DatabaseName: cfg.DatabaseName}
+		select {
+		case <-quit:
+			log.Printf("Worker goroutine received quit")
+			break loop
+		default:
+			log.Println("Waking up and looking for emails to send")
+		}
+		email, err := mongoClient.FetchReadyEmail(workerId)
+		if err == nil {
+			log.Printf("Got email %s Worker ID: %s\n", email.ID, workerId)
+			// Try to send the email
+			mc := babymailgun.MailConfig{MailHost: cfg.MailHost, MailPort: cfg.MailPort, AdminEmail: cfg.AdminEmail}
+			if err = babymailgun.SendMail(&mc, email); err != nil {
+				fmt.Println("Email sending failed ", err)
+				// TODO Set the reason on the update document
+				// TODO Some errors are definitely catastrophic
+			}
+			// Process the email response to see if any recipients failed and create
+			// and update document. Additionally, "release" the email by setting the
+			// worker_id to nil
+
+			log.Printf("Updating and releasing email %s\n", email.ID)
+			if err := mongoClient.UpdateEmail(email); err != nil {
+				// TODO Couldn't update the email, what do we do?
+				log.Println(err)
+			}
+		} else {
+			log.Printf("Error while fetching emails: %s, %t\n", err, err)
+		}
+
+		log.Printf("Going back to sleep for %d seconds\n", cfg.WorkerSleep)
+		time.Sleep(time.Duration(cfg.WorkerSleep) * time.Second)
```

Comment:
Good point. I was think it would just naively sleep, but it would be better if we actually go back to sleep if we don't find anything to work on.

Also worth pointing out that the current find ONLY worries about whether or not something has an assigned worker goroutine. It still needs logic to check for the status and whether or not N minutes (as per requirements, 10 minutes) have elapsed since the last check